### PR TITLE
Deduplicate shared params before constructing Reducer in DDP

### PR DESCRIPTION
**Pull request by mcarilli**
Currently, `torch.nn.parallel.DistributedDataParallel(model...)` doesn't deduplicate params shared across `model`'s child Modules before calling Reducer with the param list. This can cause Reducer to register more than one hook on the shared param(s), at which point who knows what happens.

We ran into this in mlperf BERT, which has at least one param shared across submodules (an embedding weight iirc, not 100% sure). Running with `gradient_as_bucket_view = False` produced different numerics from running with `gradient_as_bucket_view = True` (which i guess is one potential consequence of multiple DDP hooks on a given param, not sure why, i'd have to dig further).

This PR changes DDP to deduplicate shared params (a small diff), and adds some tests (right now just `test_ddp_weight_sharing`, but I'll add more). `test_ddp_weight_sharing` fails with bad numerics on current master (proving the shared param issue is real) and passes with the deduplication diff.

### CODE REVIEW

**Comment by zhaojuanmao** [comment]
awesome stuff! thanks for contributing this, I will take a look at it later on today!

File: test/distributed/test_c10d.py
Comment (zhaojuanmao) [inline test/distributed/test_c10d.py:4056]: just for my understanding: 2. * (world * (world + 1.) / 2.) / world = world +1, right? also, would you please help commenting why the analytic is calculated in this way?

**Review by zhaojuanmao — COMMENTED** [review]
Codes and tests look good! just one question regarding the test. will approve the diff after test_ddp_checkpointing and test_ddp_checkpointing_weight_sharing are added

File: test/distributed/test_c10d.py
Comment (mcarilli) [reply test/distributed/test_c10d.py:4056]: Do you mean, where the math comes from, or why I didn't simplify it?  I didn't simplify it because I forgot :P

Comment (mcarilli) [reply test/distributed/test_c10d.py:4056]: Do you want me to add a comment how I got the analytic value based on what each layer's backward does?

Comment (zhaojuanmao) [reply test/distributed/test_c10d.py:4056]: yes, I'm a little confused about the formula, also, feel like it can be simplified as well?

**Review by rohan-varma — COMMENTED** [review]
Wondering if it's also worth it to add a check here: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/reducer.cpp#L97, ensuring that hooks are registered at most once on a parameter?

**Review by zhaojuanmao — APPROVED** [review]
looks good, checkpoint tests on top of this PR also passed in https://github.com/pytorch/pytorch/pull/52894

File: test/distributed/test_c10d.py
Comment (zhaojuanmao) [inline test/distributed/test_c10d.py:4095]: would you please also add "@skip_if_not_multigpu"? otherwise it will fail when there is no gpus found, thanks

Comment (mcarilli) [reply test/distributed/test_c10d.py:4096]: done

File: test/distributed/test_c10d.py
Comment (mcarilli) [reply test/distributed/test_c10d.py:4123]: Added explanation. If you don't mind i'd rather keep it unsimplified, it's easier to map to the explanation that way.

Comment (zhaojuanmao) [reply test/distributed/test_c10d.py:4123]: sounds good!

**Comment by zhaojuanmao** [comment]
@mcarilli would you please rebase to latest master again? there is merge conflict..., thanks!

**Comment by zhaojuanmao** [comment]
@mcarilli looks like master was broken..., pytorchbot did not work for rebasing. would you please rebase and push it again..? thanks

**Comment by zhaojuanmao** [comment]
failed tests are not related. checked the distributed and c10d tests passed. landing

**Comment by mruberry** [comment]
Sorry @mcarilli , but I'm unlanding this (for now, anyway) because the incidence of pytorch_linux_xenial_py3_clang5_asan_test2 failures shortly after this was landed greatly increased. Will update with whether the build becomes more stable after reverting this.

**Comment by mcarilli** [comment]
Closing, resubmitted to https://github.com/pytorch/pytorch/pull/53279.

**Pull request closed without merging**
